documentation/client-side-forms/appendix-client-side-form-binding-reference.asciidoc

```diff
@@ -0,0 +1,211 @@
+
+Vaadin automatically generates Model classes for server-side endpoints from Java beans. Usually there is no need to define models manually.
+
+Models are instanciated by the `Binder`, which recieves the form model class as the `Binder` constructor argument. The nested model instances are created automatically for every field.
+
+Technically, every model instance represents either a key of a parent model, or the value of the Binder itself. The model constructors receive the following arguments:
+
+- `parent: ModelParent<T>` — the parent object reference: either the parent model instance, or the `Binder` instance
+- `key: keyof any` — the key in the parent object to get the value from / set the value to, could be `string` when the parent is an object or `number` when the parent is an array
+- `...validators: ReadonlyArray<Validator<T>>` — the default validators, typically exported from Java JSR-380 annotations
+
+=== AbstractModel
+
+All models subclass from the `AbstractModel<T>` TypeScript class, where the `T` type argument refers to the value type.
+
+AbstractModel, as well as any subclass, define a static method `static createEmptyValue(): T`, that returns the empty value of the subject model type:
+
+[source, typescript]
+----
+const emptyPerson: Person = PersonModel.createEmptyValue();
+console.log(emptyPerson); // {"fullName": ""}
+----
+
+As with any JavaScript object, `AbstractModel` has `toString(): string` and `valueOf(): T` instance methods, that are handy for template expressions:
+
+[soruce, html]
+----
+<p>
+  Full name:
+  ${model.firstName.toString()}
+  ${model.middleName.toString()}
+  ${model.lastName.toString()}
+</p>
+
+<p>
+  Cost:
+   ${model.quantity.valueOf() * model.price.valueOf()}
+</p>
+----
+
+=== Primitive models
+
+There are the builtin models that represent the common primitive field types:
+
+|===
+| Type | Value type `T` | Empty value (`Model.createEmptyValue()` result)
+
+| `StringModel` | `string` | `''`
+| `NumberModel` | `number` | `0`
+| `BooleanModel`  | `boolean` | `false`
+|===
+
+Primitive models subclass `PrimitiveModel<T>`. Primitive models are leaf, i. e., do not have nested field keys.
+
+=== Object models
+
+`ObjectModel<T>` is a common superclass for all generated bean models.
+
+There is no API added in the `ObjectModel<T>` over `AbstractModel<T>` itself.
+
+The generated subclasses of `ObjectModel<T>` define the type argument constrains and the default type, as well as list the bean fields as in Java , e. g. for the following Java bean:
+
+.IdEntity.java
+[source, java]
+----
+public class IdEntity {
+    private String idString;
+
+    public String getIdString() {
+        return idString;
+    }
+
+    public void setIdString(String idString) {
+        this.idString = idString;
+    }
+}
+----
+
+.Person.java
+[source,java]
```

Comment:
Minor consistency issue. All other blocks in this file have a space after the comma here. (looks like it's quite inconsistent in the whole repository too, but would be good to at least try to keep some kind of consistency for new content.)
```suggestion
[source, java]
```

Reply:
Yeah, it’s very inconsistent. I would prefer to have the space after the comma as well.